mautic/mautic
# Failed Event Notifications

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Y]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [N]
| Automated tests included?              | [Y] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

This is a PR for changing the way we handle notifications for failed events within campaigns.

This PR leverages the event we already fire, `CampaignEvents::ON_EVENT_FAILED`. We're moving the notification into an event listener in which we are doing some checks to see if the given campaign event has failed for more than 5% of the contacts within the campaign. If so, notify the user. If its failed more than 10%, disable the campaign and notify the user of that.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
1. Create a template email.
2. Create a segment and a campaign sourced from that segment.
3. Put 10 contacts in that segment (preferably faked ones that can actually receive the email)
4. Set your Email Settings to "Other SMTP" and give bogus credentials so that the email sending will fail.
5. In that campaign, create a "send email" action. Select the email you created earlier and choose it as marketing email.
6. Trigger the campaign and see that the emails failed.
7. You'll start generating a ton of entries in your notifications table that the action failed due to the contact having already received the marketing email.
8.you'll only start generating errors after 5% of the contacts have failed 
9. You'll see that only 1 notification is created until the campaign automatically unpublishes itself.
10. Ensure that you get an email stating your campaign has been unpublished. This you may not see as email settings are changed.
